src/Shared/Sdk/Providers/ServiceClients/StorageClientFactory.cs

```diff
@@ -0,0 +1,32 @@
+using System;
+using System.Net.Http;
+using System.Threading.Tasks;
+using ACMTTU.NoteSharing.Shared.DataContracts;
+using Microsoft.Azure.Cosmos;
+using Microsoft.Azure.Storage;
+using Microsoft.Azure.Storage.Blob;
+using Newtonsoft.Json;
+
+namespace ACMTTU.NoteSharing.Shared.SDK.Clients
+{
+    public class StorageClientFactory : ClientFactory
+    {
+        public StorageClientFactory(IHttpClientFactory factory) : base(factory)
+        {
+            this.secretServiceUrl = new Uri(this.secretServiceUrl.AbsoluteUri + "Database");
+        }
+
+        public async Task<CloudBlobClient> GetClient(string connectionString)
```

Comment:
GetClient does not override anything on the base class and therefore on any location where a ClientFactory is required GetClient cannot be called.

The fundamental problem is that the two db types are not polymorphic to one another, and therefore expressing in the type system that a method should return one or the other, when they do not share a base methods is impossible.

What we need instead is some kind of ad-hoc polymorphism OR structural polymorphism preferably in the form of type classes.

Reply:
Actually it's late and I'm overcomplicating this. This is a generic abstract factory.

Reply:
Note: This does not damage type safety in any way. Any service which needs a blob or relational service will accept in it's constructor parameter the ClientFactory<whatever> where whatever is the type of storage it needs.

I think we can make this cleaner if the ClientFactory<T>.CreateClient<T> => DBClient<T>. DBClient<T> would then contain all of the logic pertaining to using that storage mechanism. This class would service as an anti corruption layer.
This is a bad idea if we intend to use all of the mechanisms of each DB technology whenever the DbClient<T> is used.

What Would Dijkstra Do?

Reply:
OoooooOoooOoo I like the changes. But how does that I create our type safety?

Reply:
?

Reply:
Nvm i read it and figured it out